{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lab 3: Tree inference using parsimony; bootstrapping\n",
    "\n",
    "<img src=\"http://bitesizebio.s3.amazonaws.com/wp-content/uploads/2012/01/Pseudoreplication_header-image_cropped.jpg\" />\n",
    "\n",
    "### By the end of the lab, you will know how to:\n",
    "1. Reconstruct a tree using the maximum parsimony optimality criterion.\n",
    "2. Perform a bootstrap analysis to estimate confidence in the inferred topology\n",
    "3. View the results in FigTree\n",
    "\n",
    "We will use the rhodopsin and _wingless_ alignments from last week. In case those files have gone missing, they can be downloaded from the course website:\n",
    "\n",
    "- [LWRh.afa](https://sites.google.com/a/fieldmuseum.org/rtol/lab-exercises/LWRh.afa)\n",
    "- [Wg.afa](https://sites.google.com/a/fieldmuseum.org/rtol/lab-exercises/Wg.afa)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tree inference using PAUP*\n",
    "\n",
    "[__PAUP*__](http://phylosolutions.org) is one of the original phylogeny inference programs. Its original name was just PAUP, for Phylogenetic Analysis using Parsimony, but as likelihood-based methods were implemented, it was changed to PAUP* - Phylogenetic Analysis using PAUP - a [recursive acronym](https://en.wikipedia.org/wiki/Recursive_acronym) in the venerable UNIX tradition.\n",
    "\n",
    "The PAUP* program and manual can be downloaded from its website: http://phylosolutions.org. There is a GUI (mouse-driven) version for Mac computers, but we will use the command-line version, which works across operating systems and will soon be free and open-source."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<img style=\"float: right;\" src=\"http://www.reelab.net/todo.png\">\n",
    "\n",
    "### Download PAUP* (if not already installed)\n",
    "\n",
    "The Mac command-line version is here: http://phylosolutions.com/paup-test/paup4a157_osx.gz\n",
    "\n",
    "Download and extract the binary executable file (**`paup4a157_osx`**) to your working directory.\n",
    "\n",
    "Open a Terminal and, if necessary, **`cd`** to the working directory.\n",
    "\n",
    "For convenience, rename the file **`paup4a157_osx`** to just **`paup`**.\n",
    "\n",
    "  ```bash\n",
    "  mv paup4a157_osx paup\n",
    "  ```\n",
    "\n",
    "Make sure it is recognized as an executable program:\n",
    "\n",
    "  ```bash\n",
    "  chmod +x paup\n",
    "  ```\n",
    "  \n",
    "Run the program. Since it resides in the current working directory **`./`**, which is not in our [PATH environment variable], we need to specify its path:\n",
    "\n",
    "  ```bash\n",
    "  ./paup\n",
    "  ```\n",
    "\n",
    "Note how it replaces the Terminal shell prompt with its own command prompt. You can access a rudimentary help system for the available commands by typing **`?`**:\n",
    "\n",
    "  ```\n",
    "  paup> ?\n",
    "  ```\n",
    "  \n",
    "To quit the program, enter **`q`**.\n",
    "\n",
    "  ```\n",
    "  paup> q\n",
    "  ```\n",
    "\n",
    "[PATH environment variable]: https://en.wikipedia.org/wiki/PATH_(variable)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<img style=\"float: right;\" src=\"http://www.reelab.net/todo.png\">\n",
    "\n",
    "### Convert your aligned FASTA files to NEXUS files\n",
    "\n",
    "PAUP* and Mesquite require input files in [NEXUS format](https://en.wikipedia.org/wiki/Nexus_file). Like FASTA files, NEXUS files are just text files with a particular structure (and vocabulary).\n",
    "\n",
    "Last time we used MUSCLE to create FASTA alignments, **`LWRh.afa`** and **`Wg.afa`**.\n",
    "\n",
    "Open each file in Mesquite. This will prompt you save it in NEXUS format.\n",
    "\n",
    "Save each file with the default names, **`LWRh.afa.nex`** and **`Wg.afa.nex`**.\n",
    "\n",
    "Open one in a text editor (you can do this in Jupyter, from the Home tab) and inspect the NEXUS format. Note:\n",
    "\n",
    "* the leading **`#NEXUS`** token\n",
    "* the modular structure of named _blocks_, demarcated by **BEGIN _BLOCKNAME_;** and **END _BLOCKNAME_;** statements\n",
    "* comments (ignored by programs reading the file) are enclosed in square brackets: **`[this is a comment]`**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<img style=\"float: right;\" src=\"http://www.reelab.net/todo.png\">\n",
    "\n",
    "### Create and run a NEXUS file of commands to infer the _wingless_ gene tree for _Pseudomyrmex_\n",
    "\n",
    "In the Jupyter Home tab, create a new text file and name it **`run-parsimony.nex`**. Enter the following commands:\n",
    "\n",
    "```\n",
    "#NEXUS\n",
    "\n",
    "begin paup;\n",
    "log file=Wg-parsimony.log replace;\n",
    "execute Wg.afa.nex;\n",
    "outgroup 'Myrcidris_epicharis_AY703651';\n",
    "set increase=auto;\n",
    "hsearch addseq=random;\n",
    "savetrees file=Wg-mp-trees.nex root=yes brlens=yes replace;\n",
    "end;\n",
    "```\n",
    "\n",
    "Save the file and make sure it is in the working directory with your alignment files. If on lab computer, move all files to Documents where PAUP is located.\n",
    "\n",
    "Run the file in PAUP. If your PAUP session is still open in the Terminal, use the **`execute`** command:\n",
    "\n",
    "  ```\n",
    "  paup> execute run-parsimony.nex\n",
    "  ```\n",
    "  \n",
    "Otherwise, you can pass the file as an argument to the PAUP executable:\n",
    "\n",
    "  ```bash\n",
    "  ./paup run-parsimony.nex\n",
    "  ```\n",
    "  \n",
    "Look at the output, either on the Terminal screen, or in the log file **`Wg-parsimony.log`**. You have just done a bare-bones heuristic search for the most parsimonious trees for the _wingless_ gene. Congratulations!\n",
    "\n",
    "Let's examine our NEXUS command file to understand what it's doing. The first line simply declares it to be a NEXUS file. Then there is a single **`paup`** block, demarcated by the **`begin ...`** and **`end;`** statements.\n",
    "\n",
    ">***Note***: NEXUS commands are case-insensitive.\n",
    "\n",
    "Line-by-line breakdown:\n",
    "\n",
    "    log file=Wg-parsimony.log replace;\n",
    "    \n",
    "* Record everything to the specified file, and to overwrite (replace) the file if it already exists. **`replace`** is an option that can be changed: type **`help log`** for details.\n",
    "\n",
    "\n",
    "    execute Wg.afa.nex;\n",
    "    \n",
    "* Execute (read commands from) the file `Wg.afa.nex`, which in this case just contains the sequence data matrix.\n",
    "\n",
    "\n",
    "    outgroup Myrcidris_epicharis_AY703651;\n",
    "    \n",
    "*  Because we are interested in the rooted tree of _Pseudomyrmex_, we specify an outgroup, _Myrcidris_epicharis_AY703651_. We could specify the outgroup as any subset (including all) of the non-_Pseudomyrmex_ sequences, but for simplicity we'll use just the first one.\n",
    "\n",
    "\n",
    "    set increase=auto;\n",
    "\n",
    "* By default, PAUP will save only 100 most-parsimonious trees in memory at a time. This command increases this limit automatically as additional trees are found, at the risk of exhausting all the computer's memory.\n",
    "\n",
    "\n",
    "    hsearch addseq=random;\n",
    "    \n",
    "* Do the heuristic search, using random addition for each starting tree replicate. This command has many options, for building the starting tree, branch swapping, replicating the search, etc.; see **`help hsearch`** to see them.\n",
    "\n",
    "\n",
    "    savetrees file=Wg-mp-trees.nex root=yes brlens=yes replace;\n",
    "\n",
    "* Save all the most-parsimonious trees, with their parsimony branch lengths, to a NEXUS file, replacing it if necessary.\n",
    "\n",
    "***How many trees were found?***\n",
    "\n",
    "***What was their score, meaning their length - the number of nucleotide changes required to explain the data?***\n",
    "\n",
    "***Open the treefile in FigTree and page through some of the trees, to get a sense of how different they are.***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<img style=\"float: right;\" src=\"http://www.reelab.net/todo.png\">\n",
    "\n",
    "### Make a strict consensus tree\n",
    "\n",
    "Given a number of equally optimal trees, one naturally wants to know what relationships they have in common -- i.e., their ***consensus***. There are different kinds and degrees of consensus; the simplest, easiest to interpret, and arguably the most useful is the **strict consensus**. This method computes the set of clades (or in the case of unrooted trees, the bipartitions) that are common to ALL of the input trees. This set can then be summarized as a single tree, the ***strict consensus tree***.\n",
    "\n",
    "Here are the commands to do this in PAUP*. We will save the tree both as a NEXUS file and a plain Newick file.\n",
    "\n",
    "    contree / strict treefile=Wg-mp-contree.nex replace;\n",
    "    gettrees file=Wg-mp-contree.nex;\n",
    "    savetrees file=Wg-mp-contree.newick format=newick;\n",
    "\n",
    "***View the consensus tree in FigTree***. What does the number of polytomies suggest to you about how informative the _wingless_ gene is to _Pseudomyrmex_ phylogeny?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<img style=\"float: right;\" src=\"http://www.reelab.net/todo.png\">\n",
    "\n",
    "### Estimate confidence in the tree using the non-parametric bootstrap\n",
    "\n",
    "Create a new NEXUS file for a bootstrap analysis:\n",
    "\n",
    "```\n",
    "#NEXUS\n",
    "\n",
    "begin paup;\n",
    "execute Wg.afa.nex;\n",
    "outgroup 'Myrcidris_epicharis_AY703651';\n",
    "set maxtrees=1 increase=no;\n",
    "hsearch addseq=random;\n",
    "bootstrap nrep=100 keepall=yes;\n",
    "savetrees file=Wg-mp-bootcon.newick format=newick root=yes replace;\n",
    "end;\n",
    "```\n",
    "\n",
    "In a new Terminal, **`cd`** to the working directory and execute this file in PAUP*.\n",
    "\n",
    "***How quickly does the analysis finish, compared to the previous heuristic search?***\n",
    "\n",
    "The commmands tell PAUP to do _**100** replicate heuristic searches_ - compared to the previous _single_ search.\n",
    "\n",
    "***Why then did this analysis finish so quickly?***\n",
    "\n",
    "The basic output of a bootstrap analysis is a table of clades or bipartitions and their associated frequencies -- the percentage of the replicates in which the clade was found by the heuristic search. These percentages are commonly called ***bootstrap values***.\n",
    "\n",
    "By default, PAUP computes a ***majority-rule consensus tree*** from this table. The tree is composed of clades found in over 50% of the  replicate searches. It will also contain clades that are _not contradicted_ by any other tree, in order of their frequency.\n",
    "\n",
    "***Open the bootstrap consensus tree*** (the file `Wg-mp-bootcon.newick`) in a text editor and look at the Newick format:\n",
    "\n",
    "    ...((Pseudomyrmex_elongatulus_AY703659:100,Pseudomyrmex_apache_AY703652:100):55,...\n",
    "    \n",
    "The Newick convention is to write _branch lengths_ as numbers following colons. PAUP writes bootstrap consensus trees with ***bootstrap values*** as branch lengths. So here, the clade of sister species _Pseudomyrmex elongatulus_ + _P. apache_ has a bootstrap value of 55%.\n",
    "\n",
    "***Open the bootstrap consensus tree in FigTree.*** Use the side panel to show the bootstrap values:\n",
    "\n",
    "- Check and expand **Branch Labels**\n",
    "- Choose **Display: Branch lengths (raw)**\n",
    "\n",
    "***How many internal nodes have \"strongly supported\" nodes -- with bootstrap values > 70%?***\n",
    "\n",
    "FigTree will show the tree as a ***phylogram*** (if you can't remember what this means, refer to your lecture notes!) because the tree contains branch lengths. We know they aren't branch lengths, but bootstrap values. Visually, this means that longer branches are more strongly supported."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<img style=\"float: right;\" src=\"http://www.reelab.net/todo.png\">\n",
    "\n",
    "### One gene down, one to go\n",
    "\n",
    "Repeat the above analyses for the rhodopsin alignment, `LWRh.afa.nex`. Make sure to edit your executable nexus files to match the information for LWRh.afa.nex and so that it doesn't replace your existing files.\n",
    "\n",
    "***Which gene has more strongly supported clades?***\n",
    "\n",
    "***Are there any strongly supported clades in common?***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<img style=\"float: right;\" src=\"http://www.reelab.net/todo.png\">\n",
    "\n",
    "### Final exercise: a combined analysis of both genes\n",
    "\n",
    "We will repeat the exercise with the _wingless_ and rhodopsin genes laid end-to-end, i.e., ***concatenated***.\n",
    "\n",
    "\n",
    "\n",
    "***The problem:*** recall that when we created the single-gene FASTA files in Lab 1, we labeled the sequences like this:\n",
    "\n",
    "    >Myrcidris_epicharis_AY703651\n",
    "    GTTGCCGAACTTTCGCGTGGTTGGAGACAATCTGAAAGATCGCTTCGATGGAGCATCCCG\n",
    "    ...\n",
    "\n",
    "That is, we _included the GenBank accession number_ in the sequence label. This seemed like a good idea at the time, since it kept the provenance of the data in the file. But now we want to ***concatenate*** the alignments, and make a \"supermatrix\" where the row labels contain only the species name.\n",
    "\n",
    "The script below concatenates the alignments and saves the result as a new FASTA file.\n",
    "\n",
    "***Create a new code cell below and execute the script.***\n",
    "\n",
    "```python\n",
    "from Bio import AlignIO\n",
    "\n",
    "# A convenience function for stripping the accession number\n",
    "def stripacc(label):\n",
    "    # We know the label is of the form: Genus_species_accession\n",
    "    # so we split it on underscores and rejoin it, minus the last element\n",
    "    return '_'.join(label.split('_')[:-1])\n",
    "\n",
    "# Let's read in the rhodopsin alignment first\n",
    "fn = 'LWRh.afa'\n",
    "aln = AlignIO.read(fn, format='fasta')\n",
    "print('{} has length {}'.format(fn, aln.get_alignment_length()))\n",
    "\n",
    "# Strip the accession number from each Bio.SeqRecord in the alignment\n",
    "for s in aln:\n",
    "    s.id = stripacc(s.id)\n",
    "    s.description = ''\n",
    "\n",
    "# Make a dictionary for looking up SeqRecords by id\n",
    "d = dict([ (s.id, s) for s in aln ])\n",
    "\n",
    "# Now read in the wingless alignment\n",
    "fn = 'Wg.afa'\n",
    "wg = AlignIO.read(fn, format='fasta')\n",
    "print('{} has length {}'.format(fn, wg.get_alignment_length()))\n",
    "\n",
    "# Concatenate the sequences. Every wingless sequence has one,\n",
    "# and only one, matching rhodopsin sequence, making things simple\n",
    "for s in wg:\n",
    "    k = stripacc(s.id)\n",
    "    lwrh = d[k]  # matching rhodopsin sequence\n",
    "    lwrh.seq += s.seq  # this 'adds' the sequences together, i.e., concatenates them\n",
    "\n",
    "print('combined alignment has length {}'.format(aln.get_alignment_length()))\n",
    "\n",
    "AlignIO.write(aln, 'LWRh.Wg.afa', format='fasta')\n",
    "```\n",
    "\n",
    "***Do a bootstrap analysis of the combined matrix. How do the support values compare to the individual gene analyses?***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Optional: interactive tree comparison\n",
    "\n",
    "Few GUI tools exist for interactively comparing trees. One is http://phylo.io/. Another is implemented in a Python package, [**`ivy`**](https://github.com/rhr/ivy), written by yours truly (RR). The code is not well documented, nor extensively tested, and contains bugs. With that in mind, I've written a tutorial on how to use **`ivy`** to visually and interactively compare trees such as the bootstrap consensus trees computed above. Please try it out if you have time: download the notebook below and run it from this working directory.\n",
    "\n",
    "https://www.dropbox.com/s/etsaf6lpwr0jcl8/tree-compare.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<img src=\"http://3.bp.blogspot.com/-F_kwLarXHcc/Ux5fciXTyYI/AAAAAAAANNY/3MrB_L23WxQ/s1600/Porky_Pig_Thats_All_Folks.jpg\" />"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
